File: SEGUNDO PERIODO/AEDS II/TPS/01/Palindromo.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

// função recursiva para verificar se é palíndromo
bool Palindromo(int i, int f, char w[])
{
    if (i >= f)
    {
        return true; // condição de parada: se os índices se cruzaram, é palíndromo
    }
    if (w[i] != w[f])
    {
        return false; // se encontrou um caractere diferente, não é palíndromo
    }
    // chamada recursiva avançando o início e recuando o fim
    return Palindromo(i + 1, f - 1, w);
}

int main(void)
{
    char word[1001];

    // lê linhas até o EOF com fgets
    while (fgets(word, sizeof(word), stdin))
    {
        // remove \n (ou \r\n) do final da string
        word[strcspn(word, "\r\n")] = '\0';

        // condição de parada: se a palavra for "FIM", encerra o programa
        if (strcmp(word, "FIM") == 0)
        {
            break;
        }

        int inicio = 0, fim = (int)strlen(word) - 1; // declara um início e um fim para a comparação

        bool ok = Palindromo(inicio, fim, word);

        if (ok)
            printf("SIM\n");
        else
            printf("NAO\n");
    }

    return 0;
}

File: SEGUNDO PERIODO/AEDS II/TPS/01/Palindromo.java
import java.util.Scanner;

public class Palindromo {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in, "UTF-8");

        while (sc.hasNextLine()) {
            String word = sc.nextLine();

            // condição de parada: se a palavra for "FIM", o programa encerra
            if (word.equalsIgnoreCase("FIM")) {
                break;
            }

            int i = 0, j = word.length() - 1; // declara um inicio e um final para a comparação

            while (i < j && word.charAt(i) == word.charAt(j)) {
                // condição : se o i < j e as letras da string de acordo com a posição i e j são
                // iguais
                // o while irá continuar funcionando ate q pare;
                i++;
                j--;
            }

            if (i < j)
                System.out.println("NAO"); // quer dizer que saiu do while antes de i ser >= a j;
            else
                System.out.println("SIM"); // while continuo ate q o i ficou com o valor de >= a j;
        }

        sc.close();
    }
}
